### PR TITLE
feat: Add PyTorch Lightning zero-code integration

### LINKED ISSUE #23: Test TraceML with PyTorch Lightning Trainer (no Lightning-specific integration)
TraceML is designed to instrument PyTorch training loops with step-level semantics and low-overhead system signals (CUDA events, memory, dataloader time).

Before adding any Lightning-specific adapters, we want to answer a basic question:

_Does TraceML work with PyTorch Lightning “as-is”?_

This issue is about default compatibility only, without callbacks, hooks, or code changes.

### PR DESCRIPTION
**Pull request by ppraneth**
## Summary

Adds automatic PyTorch Lightning integration to TraceML. Users can now get full training observability with zero additional code when using Lightning (Closes #23 ).

image

## Features

- **Zero-code integration**: TraceML automatically injects a callback into Lightning's `Trainer` that captures step-level metrics
- **Step timers**: Forward, backward, and optimizer_step timing captured automatically via Lightning hooks
- **Per-layer instrumentation**: Works with [trace_model_instance()](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/decorators.py:89:0-159:80) for deep-dive memory and timing analysis
- **No manual [trace_step()](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/decorators.py:80:0-87:36) needed**: Lightning's callback lifecycle handles step boundaries

### Core Implementation
- **[src/traceml/utils/lightning_patch.py](cci:7://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:0:0-0:0)**: New [TraceMLCallback](cci:2://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:18:4-84:42) that hooks into:
  - [on_train_batch_start](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:32:8-38:69) / [on_train_batch_end](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:68:8-84:42) - step boundaries
  - [on_before_backward](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:40:8-47:71) / [on_after_backward](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:49:8-54:49) - backward timing
  - [on_before_optimizer_step](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:56:8-59:78) / [on_before_zero_grad](cci:1://file:///d:/Python/traceopt-fork/traceml/src/traceml/utils/lightning_patch.py:61:8-66:50) - optimizer timing
  - Forward timing captured between batch_start and before_backward

### Example
- **[src/examples/test_lightning_traceml.py](cci:7://file:///d:/Python/traceopt-fork/traceml/src/examples/test_lightning_traceml.py:0:0-0:0)**: New example demonstrating Lightning integration (mirrors [tracing_on_cnn.py](cci:7://file:///d:/Python/traceopt-fork/traceml/src/examples/tracing_on_cnn.py:0:0-0:0))

### Documentation
- **[README.md](cci:7://file:///d:/Python/traceopt-fork/traceml/README.md:0:0-0:0)**: 
  - Added Lightning installation instructions
  - New "PyTorch Lightning (zero-code integration)" section with example code
  - Updated roadmap to show Lightning as ✅ Done

### Dependencies
- **[pyproject.toml](cci:7://file:///d:/Python/traceopt-fork/traceml/pyproject.toml:0:0-0:0)**: Added `lightning>=2.6.0` as optional dependency
  - Install with: `pip install traceml-ai[lightning]`

### CODE REVIEW

**Comment by ppraneth** [comment]
@abhinavsriva Can you take a look at this pr?

**Comment by abhinavsriva** [comment]
@ppraneth can you please resolve conflicts here and see if it still works, otherwise you can use wrapper similar like Hugging face trainer.

File: src/traceml/decorators.py [hunk 1]
```diff
@@ -48,6 +48,14 @@
 patch_backward()
 
 
+try:
```
Comment (abhinavsriva) [inline src/traceml/decorators.py]: I think monkey-patching won't be good for lightning users, we can use official callbacks.

**Comment by ppraneth** [comment]
**Regarding Comments 1 & 3 (Monkey-Patching & Callbacks):**
 I have removed the `patch_lightning()` injection logic and the `Trainer.__init__` monkey-patch entirely. The integration is now purely an official callback. Users can now simply import and use it cleanly:
`trainer = L.Trainer(callbacks=[TraceMLCallback()])`

I have updated the README and example files to reflect this manual usage API.

**Regarding Comment 2 (Gradient Accumulation):**
I have verified by tracing the execution hooks, PyTorch Lightning triggers `on_train_batch_end` for every single micro-batch during accumulation, but skips `on_before_optimizer_step`. 

If we were to group them all into one single macro-step, it might breaks TraceML's UI rendering. TraceML’s step-combination logic relies on `forward_time`, `backward_time`, and `optimizer_step` belonging to the exact same step number to accurately compute the “Wait” proxy. If we only emit an optimizer step every *N* micro-batches, the arrays become misaligned and the dashboard errors out. 

To solve this cleanly:
I kept the logic that advances the TraceML step counter for *every micro-batch*. This actually provides better, more granular observability (you can see the forward/backward time for each specific micro-batch). Then, to keep the dashboard’s step alignment intact, the callback now emits a dummy `0ms` duration optimizer event for accumulation steps where the optimizer doesn’t actually run.

Let me know what you think of this approach?
